src/osdc/Objecter.h

```diff
@@ -2192,12 +2348,79 @@ class Objecter : public md_config_obs_t, public Dispatcher {
   void set_client_incarnation(int inc) { client_inc = inc; }
 
   bool have_map(epoch_t epoch);
-  /// wait for epoch; true if we already have it
-  bool wait_for_map(epoch_t epoch, Context *c, int err=0);
-  void _wait_for_new_map(Context *c, epoch_t epoch, int err=0);
-  void wait_for_latest_osdmap(Context *fin);
-  void get_latest_version(epoch_t oldest, epoch_t neweset, Context *fin);
-  void _get_latest_version(epoch_t oldest, epoch_t neweset, Context *fin);
+
+  struct CB_Objecter_GetVersion {
+    Objecter *objecter;
+    std::unique_ptr<OpCompletion> fin;
+
+    CB_Objecter_GetVersion(Objecter *o, std::unique_ptr<OpCompletion> c)
+      : objecter(o), fin(std::move(c)) {}
+    void operator()(boost::system::error_code ec, version_t newest,
+		    version_t oldest) {
+      if (ec == boost::system::errc::resource_unavailable_try_again) {
+	// try again as instructed
+	objecter->wait_for_latest_osdmap(std::move(fin));
+      } else if (ec) {
+	ceph::async::post(std::move(fin), ec);
+      } else {
+	auto l = std::unique_lock(objecter->rwlock);
+	objecter->_get_latest_version(oldest, newest, std::move(fin));
+      }
+    }
+  };
+
+  template<typename CompletionToken>
+  typename boost::asio::async_result<CompletionToken, OpSignature>::return_type
+  wait_for_map(epoch_t epoch, CompletionToken&& token) {
+    boost::asio::async_completion<CompletionToken, OpSignature> init(token);
+
+    if (osdmap->get_epoch() >= epoch) {
+      boost::asio::post(service,
+			ceph::async::bind_handler(std::move(init.completion_handler),
+						  boost::system::error_code()));
+    } else {
+      monc->get_version("osdmap",
+			CB_Objecter_GetVersion(
+			  this,
+			  OpCompletion::create(service.get_executor(),
+					       std::move(init.completion_handler))));
+    }
+    return init.result.get();
+  }
+
+  void _wait_for_new_map(std::unique_ptr<OpCompletion>, epoch_t epoch,
+			 boost::system::error_code = {});
+
+  template<typename CompletionToken>
+  typename boost::asio::async_result<CompletionToken, OpSignature>::return_type
+  wait_for_latest_osdmap(CompletionToken&& token) {
+    boost::asio::async_completion<CompletionToken, OpSignature> init(token);
+
+    monc->get_version("osdmap",
+		      CB_Objecter_GetVersion(
+			this,
+			OpCompletion::create(service.get_executor(),
+					     std::move(init.completion_handler))));
+    return init.result.get();
+  }
+
+  void wait_for_latest_osdmap(std::unique_ptr<OpCompletion> c) {
+    monc->get_version("osdmap",
+		      CB_Objecter_GetVersion(this, std::move(c)));
+  }
+
+  template<typename CompletionToken>
+  auto get_latest_version(epoch_t oldest, epoch_t newest,
+			    CompletionToken&& token) {
+    unique_lock wl(rwlock);
```

Comment:
the scope of this lock is problematic for some async_completion types - for example, a coroutine yield_context will cause the call to init.result.get() to suspend until the completion fires, which means that this lock will be held for the entirety of the operation

Reply:
Thought I had fixed all these but there was one left over. Fixed.